File: modules/nitro-tts/android/src/main/java/com/margelo/nitro/nitrotts/HybridTtsSession.kt
package com.margelo.nitro.nitrotts

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.Promise

@Keep
@DoNotStrip
final class HybridTtsSession : HybridTtsSessionSpec() {
    override fun load(
        paragraphs: Array<TtsParagraph>,
        initialIndex: Double,
        metadata: TtsMetadata,
        settings: TtsSettings,
    ): Promise<Unit> {
        return MainThreadPromise.run {
            TtsPlaybackStore.load(
                paragraphs,
                initialIndex.toInt(),
                metadata,
                settings,
            )
        }
    }

    override fun play(): Promise<Unit> =
        MainThreadPromise.run(TtsPlaybackStore::play)

    override fun pause(): Promise<Unit> =
        MainThreadPromise.run(TtsPlaybackStore::pause)

    override fun stop(): Promise<Unit> =
        MainThreadPromise.run(TtsPlaybackStore::stop)

    override fun skipPrevious(): Promise<Unit> =
        MainThreadPromise.run(TtsPlaybackStore::skipPrevious)

    override fun skipNext(): Promise<Unit> =
        MainThreadPromise.run(TtsPlaybackStore::skipNext)

    override fun replayCurrent(): Promise<Unit> =
        MainThreadPromise.run(TtsPlaybackStore::replayCurrent)

    override fun seekTo(index: Double): Promise<Unit> =
        MainThreadPromise.run { TtsPlaybackStore.seekTo(index.toInt()) }

    override fun updateSettings(settings: TtsSettings): Promise<Unit> =
        MainThreadPromise.run { TtsPlaybackStore.updateSettings(settings) }

    override fun addOnStateChangedListener(
        listener: (state: TtsPlaybackState) -> Unit,
    ): ListenerSubscription {
        return ListenerSubscription(TtsPlaybackStore.addStateListener(listener))
    }

    override fun addOnProgressChangedListener(
        listener: (progress: TtsProgress) -> Unit,
    ): ListenerSubscription {
        return ListenerSubscription(TtsPlaybackStore.addProgressListener(listener))
    }

    override fun addOnErrorListener(
        listener: (message: String) -> Unit,
    ): ListenerSubscription {
        return ListenerSubscription(TtsPlaybackStore.addErrorListener(listener))
    }

    override fun addOnWordRangeChangedListener(
        listener: (range: TtsWordRange) -> Unit,
    ): ListenerSubscription {
        return ListenerSubscription(TtsPlaybackStore.addWordRangeListener(listener))
    }
}

File: modules/nitro-tts/ios/HybridTtsSession.swift
import NitroModules

final class HybridTtsSession: HybridTtsSessionSpec {
  private let coordinator = TtsPlaybackCoordinator.shared

  func load(
    paragraphs: [TtsParagraph],
    initialIndex: Double,
    metadata: TtsMetadata,
    settings: TtsSettings
  ) throws -> Promise<Void> {
    return MainQueuePromise.run {
      try self.coordinator.load(
        paragraphs: paragraphs,
        initialIndex: Int(initialIndex),
        metadata: metadata,
        settings: settings
      )
    }
  }

  func play() throws -> Promise<Void> {
    return MainQueuePromise.run(coordinator.play)
  }

  func pause() throws -> Promise<Void> {
    return MainQueuePromise.run(coordinator.pause)
  }

  func stop() throws -> Promise<Void> {
    return MainQueuePromise.run(coordinator.stop)
  }

  func skipPrevious() throws -> Promise<Void> {
    return MainQueuePromise.run(coordinator.skipPrevious)
  }

  func skipNext() throws -> Promise<Void> {
    return MainQueuePromise.run(coordinator.skipNext)
  }

  func replayCurrent() throws -> Promise<Void> {
    return MainQueuePromise.run(coordinator.replayCurrent)
  }

  func seekTo(index: Double) throws -> Promise<Void> {
    return MainQueuePromise.run {
      self.coordinator.seekTo(index: Int(index))
    }
  }

  func updateSettings(settings: TtsSettings) throws -> Promise<Void> {
    return MainQueuePromise.run {
      self.coordinator.updateSettings(settings)
    }
  }

  func addOnStateChangedListener(
    listener: @escaping (TtsPlaybackState) -> Void
  ) throws -> ListenerSubscription {
    return ListenerSubscription(remove: coordinator.addStateListener(listener))
  }

  func addOnProgressChangedListener(
    listener: @escaping (TtsProgress) -> Void
  ) throws -> ListenerSubscription {
    return ListenerSubscription(remove: coordinator.addProgressListener(listener))
  }

  func addOnErrorListener(
    listener: @escaping (String) -> Void
  ) throws -> ListenerSubscription {
    return ListenerSubscription(remove: coordinator.addErrorListener(listener))
  }

  func addOnWordRangeChangedListener(
    listener: @escaping (TtsWordRange) -> Void
  ) throws -> ListenerSubscription {
    return ListenerSubscription(remove: coordinator.addWordRangeListener(listener))
  }
}
